src/durmod/kaldi-durmod.cc

```diff
@@ -0,0 +1,466 @@
+  ComputationRequest request;
+  bool need_backprop = false, store_stats = false;
+  GetComputationRequest(model_.GetNnet(), eg, need_backprop,
+                        store_stats, &request);
+  const NnetComputation &computation = *(compiler_.Compile(request));
+  NnetComputeOptions options;
+  //  if (GetVerboseLevel() >= 3)
+  //    options.debug = true;
+  NnetComputer computer(options, computation, model_.GetNnet(), NULL);
+  computer.AcceptInputs(model_.GetNnet(), eg.io);
+  computer.Forward();
+  const CuMatrixBase<BaseFloat> &nnet_output = computer.GetOutput("output");
+  output->Resize(nnet_output.NumRows(), nnet_output.NumCols());
+  nnet_output.CopyToMat(output);
+}
+
+BaseFloat NnetPhoneDurationScoreComputer::GetLogProb(
+               const std::vector<std::pair<int32, int32> > &phone_dur_context) {
+  KALDI_ASSERT(phone_dur_context.size() == model_.FullContextSize());
+  NnetExample eg;
+  MakeNnetExample(feature_maker_, phone_dur_context, model_.LeftContext(), &eg);
+  Matrix<BaseFloat> output;
+  ComputeOutputForExample(eg, &output);
+  int32 phone_duration = phone_dur_context[model_.LeftContext()].second;
+  // Please refer to MakeNnetExample() to see what the output nodes of the
+  // network show.
+  int32 duration_id = (phone_duration > model_.MaxDuration()) ?
+                                                (model_.MaxDuration() - 1):
+                                                (phone_duration - 1);
+
+  // Now we estimate the probabilities for the durations longer than
+  // max_duration. To do this we distribute the probability mass at the last
+  // node (i.e. the probability for duration==max_duration) to all the durations
+  // equal to and longer than max_duration. We assume a geometric form for this
+  // distribution and scale the probabilities such that the whole distribution
+  // sums to 1. So assuming the probability mass at duration==max_duration is
+  // P, then the probabilities for some duration >= max_duration will be
+  // P/norm_sum * alpha^(duration-max_duration+1) where norm_sum is
+  // equal to alpha/(1-alpha). We set alpha to exp(-1.0/max_duration) so
+  // that the probabilities do not decline too rapidly.
+
+  int32 actual_duration_id = phone_duration - 1;  // in case max duration was
+                                                  // infinity (i.e. we
+                                                  // had infinitely many nodes
+                                                  // at the output of the
+                                                  // network)
+  BaseFloat logprob = output(0, duration_id);
+
+  // make sure the distribution over all durations (1 to inf) sums to 1
+  BaseFloat alpha = Exp(-1.0f / model_.MaxDuration());
+  BaseFloat probability_normalization_sum = alpha / (1 - alpha);
+  if (phone_duration >= model_.MaxDuration())
+    logprob += Log(pow(alpha, actual_duration_id - duration_id + 1) /
```

Comment:
it probably doesn't affect things in practice, but please reorganize this code to avoid the possibility of underflow-- i.e. get rid of the 'pow' and do that in the log domain.